File: cpp/expert_trading_features/04_object_pool.cpp
// ============================================================================
// 04 — Object pool / free list: zero allocation on the hot path
// Compile:  clang++ -std=c++20 -O2 04_object_pool.cpp -o bin/04 && bin/04
// ----------------------------------------------------------------------------
// `new`/`malloc` are forbidden on a trading hot path: they take a lock, can
// touch the kernel, and have a long, unpredictable tail latency (page faults,
// arena contention). Instead, pre-allocate ALL objects up front and hand them
// out from an intrusive free list.
//
// Techniques shown:
//   * Contiguous storage reserved once at startup (cache-friendly, no heap churn).
//   * Intrusive free list: a freed slot stores the index of the next free slot,
//     so acquire()/release() are O(1) pointer flips with no metadata allocation.
//   * Deterministic latency: every acquire is the same handful of instructions.
// ============================================================================

#include <cstdint>
#include <vector>
#include <array>
#include <iostream>
#include <format>
#include <chrono>
#include <cassert>

// A resting order in the book — what we want to allocate millions of times/day.
struct Order {
    std::uint64_t id;
    std::int64_t  price;     // fixed-point
    std::uint32_t qty;
    char          side;      // 'B' or 'S'
};

template <typename T, std::size_t N>
class ObjectPool {
    // union-like storage: a slot is either a live T or a "next free index".
    struct Slot {
        alignas(T) unsigned char storage[sizeof(T)];
        std::size_t next_free;
    };
    std::vector<Slot> slots_;            // one big contiguous block, allocated once
    std::size_t free_head_;
    std::size_t in_use_ = 0;
    static constexpr std::size_t kNil = ~std::size_t{0};

public:
    ObjectPool() : slots_(N) {
        for (std::size_t i = 0; i < N; ++i) slots_[i].next_free = i + 1;
        slots_[N - 1].next_free = kNil;
        free_head_ = 0;
    }

    // O(1), no heap. Returns nullptr when exhausted (caller decides policy).
    template <typename... Args>
    T* acquire(Args&&... args) {
        if (free_head_ == kNil) return nullptr;
        std::size_t idx = free_head_;
        free_head_ = slots_[idx].next_free;
        ++in_use_;
        return new (slots_[idx].storage) T{std::forward<Args>(args)...};  // placement new
    }

    // O(1) return to the pool.
    void release(T* p) {
        p->~T();
        auto* slot = reinterpret_cast<Slot*>(reinterpret_cast<unsigned char*>(p));
        std::size_t idx = static_cast<std::size_t>(slot - slots_.data());
        assert(idx < N);
        slots_[idx].next_free = free_head_;
        free_head_ = idx;
        --in_use_;
    }

    std::size_t in_use() const { return in_use_; }
    static constexpr std::size_t capacity() { return N; }
};

int main() {
    static ObjectPool<Order, 1 << 16> pool;     // 65,536 orders, reserved at startup
    std::cout << std::format("pool capacity = {} orders, allocated once up front\n",
                             pool.capacity());

    // Simulate a churn of resting orders: acquire, then cancel half.
    std::array<Order*, 1000> live{};
    for (std::size_t i = 0; i < live.size(); ++i)
        live[i] = pool.acquire(Order{1000 + i, 100'0000 + static_cast<std::int64_t>(i),
                                     100, (i & 1) ? 'B' : 'S'});

    std::cout << std::format("after 1000 acquires: in_use={}\n", pool.in_use());

    std::uint64_t notional = 0;
    for (std::size_t i = 0; i < live.size(); i += 2) {   // cancel the even ones
        notional += static_cast<std::uint64_t>(live[i]->price) * live[i]->qty;
        pool.release(live[i]);
        live[i] = nullptr;
    }
    std::cout << std::format("after cancelling 500: in_use={} (slots recycled, no free())\n",
                             pool.in_use());

    // Reacquire — the freed slots come right back, hot in cache.
    for (std::size_t i = 0; i < live.size(); i += 2)
        live[i] = pool.acquire(Order{9000 + i, 200'0000, 50, 'S'});
    std::cout << std::format("after reacquiring 500: in_use={}, cancelled notional={}\n",
                             pool.in_use(), notional);

    // --- Benchmark: pool acquire/release vs new/delete ------------------------
    // The hot-path pattern is acquire -> use -> release, repeated. We time an
    // identical acquire/touch/release loop against `new`/`delete`. The pool wins
    // not just on the mean but — the part that matters for trading — by removing
    // malloc's unbounded tail (locks, arena contention, page faults).
    //
    // NOTE: we force the pointer to "escape" each iteration. Without this, clang
    // happily proves the new/delete pair is dead and deletes it outright (heap
    // allocation elision), making the heap path look instant and the bench a lie.
    static ObjectPool<Order, 1 << 16> bp;     // its own pool for the bench
    constexpr std::size_t N = 20'000'000;
    std::uint64_t sink = 0;
    auto escape = [](void* p) { asm volatile("" : : "g"(p) : "memory"); };

    auto t0 = std::chrono::steady_clock::now();
    for (std::size_t i = 0; i < N; ++i) {
        Order* o = bp.acquire(Order{i, 100'0000 + static_cast<std::int64_t>(i & 1023),
                                    100, (i & 1) ? 'B' : 'S'});
        escape(o);
        sink += o->id ^ static_cast<std::uint64_t>(o->price);   // touch it
        bp.release(o);
    }
    auto t1 = std::chrono::steady_clock::now();

    for (std::size_t i = 0; i < N; ++i) {
        Order* o = new Order{i, 100'0000 + static_cast<std::int64_t>(i & 1023),
                             100, (i & 1) ? 'B' : 'S'};
        escape(o);
        sink += o->id ^ static_cast<std::uint64_t>(o->price);
        delete o;
    }
    auto t2 = std::chrono::steady_clock::now();

    double ns_pool = std::chrono::duration<double, std::nano>(t1 - t0).count() / N;
    double ns_heap = std::chrono::duration<double, std::nano>(t2 - t1).count() / N;
    std::cout << std::format(
        "\nbench acquire/touch/release over {} iters (sink={:#x}):\n"
        "  object pool : {:.3f} ns/op\n"
        "  new/delete  : {:.3f} ns/op  (speedup {:.2f}x, and no allocator tail)\n",
        N, sink, ns_pool, ns_heap, ns_heap / ns_pool);
    return 0;
}

File: rust/expert_trading_features/examples/04_object_pool.rs
// ============================================================================
// 04 — Object pool / free list: zero allocation on the hot path
// Run:  cargo run --release --example 04_object_pool
// C++ counterpart: cpp/expert_trading_features/04_object_pool.cpp
// ----------------------------------------------------------------------------
// `Box::new` / the global allocator are forbidden on a trading hot path: they can
// lock, touch the kernel, and have a long, unpredictable tail (page faults, arena
// contention). Instead pre-allocate ALL objects up front and hand them out from
// an intrusive free list — O(1), no allocator, deterministic latency.
//
// What Rust changes vs C++:
//   * We hand out a `Handle` (a typed index), not a raw pointer. release() takes
//     the handle by value, so it's CONSUMED — the borrow/ownership rules make a
//     use-after-free or double-free far harder to write than with raw `T*`.
//   * Storage is `Vec<MaybeUninit<T>>`; the unsafe is confined to get()/release().
//   * No placement-new ceremony: `MaybeUninit::write` is the safe-ish primitive.
// ============================================================================

use std::mem::MaybeUninit;

// A resting order — what we allocate millions of times a day.
#[derive(Clone, Copy, Debug, PartialEq)]
struct Order {
    id: u64,
    price: i64,
    qty: u32,
    side: u8, // b'B' / b'S'
}

/// A typed handle into the pool. Carrying the index (not a pointer) means a stale
/// handle can be cheaply range-checked, and `release` consuming it by value stops
/// you using the slot afterwards.
#[derive(Clone, Copy, PartialEq, Eq, Debug)]
struct Handle(usize);

struct Pool<T> {
    slots: Vec<MaybeUninit<T>>,
    next_free: Vec<usize>, // intrusive free list: next_free[i] = next free slot
    free_head: usize,
    in_use: usize,
}

const NIL: usize = usize::MAX;

impl<T> Pool<T> {
    fn with_capacity(n: usize) -> Self {
        let mut slots = Vec::with_capacity(n);
        let mut next_free = Vec::with_capacity(n);
        for i in 0..n {
            slots.push(MaybeUninit::uninit());
            next_free.push(if i + 1 < n { i + 1 } else { NIL });
        }
        Pool {
            slots,
            next_free,
            free_head: if n == 0 { NIL } else { 0 },
            in_use: 0,
        }
    }

    /// O(1), no heap. Returns None when exhausted (caller decides policy).
    fn acquire(&mut self, value: T) -> Option<Handle> {
        if self.free_head == NIL {
            return None;
        }
        let idx = self.free_head;
        self.free_head = self.next_free[idx];
        self.slots[idx].write(value); // initialize the slot
        self.in_use += 1;
        Some(Handle(idx))
    }

    /// O(1) return to the pool. Consumes the handle so it can't be reused.
    fn release(&mut self, h: Handle) {
        // SAFETY: a live Handle always points at an initialized slot; we drop the
        // value exactly once and thread the slot back onto the free list.
        unsafe { self.slots[h.0].assume_init_drop() };
        self.next_free[h.0] = self.free_head;
        self.free_head = h.0;
        self.in_use -= 1;
    }

    fn get(&self, h: Handle) -> &T {
        // SAFETY: a live Handle points at an initialized slot.
        unsafe { self.slots[h.0].assume_init_ref() }
    }

    fn in_use(&self) -> usize {
        self.in_use
    }
    fn capacity(&self) -> usize {
        self.slots.len()
    }
}

fn main() {
    let mut pool: Pool<Order> = Pool::with_capacity(1 << 16);
    println!(
        "pool capacity = {} orders, allocated once up front",
        pool.capacity()
    );

    // Acquire a churn of resting orders.
    let mut live: Vec<Handle> = Vec::new();
    for i in 0..1000u64 {
        let h = pool
            .acquire(Order {
                id: 1000 + i,
                price: 1_000_000 + i as i64,
                qty: 100,
                side: if i & 1 == 0 { b'S' } else { b'B' },
            })
            .unwrap();
        live.push(h);
    }
    println!("after 1000 acquires: in_use={}", pool.in_use());

    // Cancel the even ones, recording notional.
    let mut notional: u128 = 0;
    for i in (0..live.len()).step_by(2) {
        let o = pool.get(live[i]);
        notional += o.price as u128 * o.qty as u128;
        pool.release(live[i]);
    }
    println!(
        "after cancelling 500: in_use={} (slots recycled, no allocator call)",
        pool.in_use()
    );

    // Reacquire — the freed slots come right back, hot in cache.
    for i in (0..live.len()).step_by(2) {
        live[i] = pool
            .acquire(Order {
                id: 9000 + i as u64,
                price: 2_000_000,
                qty: 50,
                side: b'S',
            })
            .unwrap();
    }
    println!(
        "after reacquiring 500: in_use={}, cancelled notional={notional}",
        pool.in_use()
    );
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn acquire_release_recycles_slots() {
        let mut p: Pool<Order> = Pool::with_capacity(2);
        let h1 = p
            .acquire(Order {
                id: 1,
                price: 100,
                qty: 1,
                side: b'B',
            })
            .unwrap();
        let h2 = p
            .acquire(Order {
                id: 2,
                price: 200,
                qty: 2,
                side: b'S',
            })
            .unwrap();
        assert_eq!(
            p.acquire(Order {
                id: 3,
                price: 300,
                qty: 3,
                side: b'B'
            }),
            None
        ); // full
        assert_eq!(p.in_use(), 2);
        p.release(h1);
        let h3 = p
            .acquire(Order {
                id: 3,
                price: 300,
                qty: 3,
                side: b'B',
            })
            .unwrap();
        assert_eq!(h3, h1); // recycled the freed slot
        assert_eq!(p.get(h2).id, 2);
        assert_eq!(p.get(h3).id, 3);
    }

    #[test]
    fn empty_pool_yields_nothing() {
        let mut p: Pool<u32> = Pool::with_capacity(0);
        assert_eq!(p.acquire(1), None);
    }
}
